### PR TITLE
Split authority

### PR DESCRIPTION
**Pull request by antrix1989**

### CODE REVIEW

File: IdentityCore/IdentityCore.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -9,13 +9,21 @@
 /* Begin PBXBuildFile section */
 		04930F851FEC8E6800FC4DCD /* MSIDAadAuthorityCache.m in Sources */ = {isa = PBXBuildFile; fileRef = 04930F831FEC8E6800FC4DCD /* MSIDAadAuthorityCache.m */; };
 		04930F8B1FEC8EB900FC4DCD /* MSIDAadAuthorityCache.m in Sources */ = {isa = PBXBuildFile; fileRef = 04930F831FEC8E6800FC4DCD /* MSIDAadAuthorityCache.m */; };
-		04930F8C1FEC8EC200FC4DCD /* MSIDAadAuthorityCache+TestUtil.m in Sources */ = {isa = PBXBuildFile; fileRef = 04930F881FEC8E9400FC4DCD /* MSIDAadAuthorityCache+TestUtil.m */; };
-		04930F8D1FEC8EC200FC4DCD /* MSIDAadAuthorityCache+TestUtil.m in Sources */ = {isa = PBXBuildFile; fileRef = 04930F881FEC8E9400FC4DCD /* MSIDAadAuthorityCache+TestUtil.m */; };
+		04930F8C1FEC8EC200FC4DCD /* (null) in Sources */ = {isa = PBXBuildFile; };
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/IdentityCore.xcodeproj/project.pbxproj]: fixed

File: IdentityCore/src/MSIDAADNetworkConfiguration.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+@property (class, nullable) MSIDAADNetworkConfiguration *defaultConfiguration;
+
+@property (nonatomic, nonnull) id<MSIDAADEndpointProviding> endpointProvider;
+
+@property (nonatomic, nullable) NSString *aadApiVersion;
+
+@property (nonatomic, nullable) NSString *aadAuthorityDiscoveryApiVersion;
+
+@property (nonatomic, nullable) NSString *drsDiscoveryApiVersion;
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/MSIDAADNetworkConfiguration.h]: we don't use it for AAD, it is for ADFS authority validation.

File: IdentityCore/src/MSIDADFSType.h [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+typedef NS_ENUM(NSInteger, MSIDADFSType)
+{
+    MSIDADFSTypeOnPrems,
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/MSIDADFSType.h]: done

File: IdentityCore/src/MSIDADFSType.h [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    MSIDADFSTypeCloud
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/MSIDADFSType.h]: done

File: IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m [hunk 5]
```diff
@@ -191,7 +192,10 @@ - (void)webView:(WKWebView *)webView decidePolicyForNavigationAction:(WKNavigati
 {
     NSURL *requestURL = navigationAction.request.URL;
     
-    MSID_LOG_VERBOSE(self.context, @"-decidePolicyForNavigationAction host: %@", [MSIDAuthority isKnownHost:requestURL] ? requestURL.host : @"unknown host");
+//    __auto_type authorityFactory = [MSIDAuthorityFactory new];
+//    __auto_type authority = [authorityFactory authorityFromUrl:account.authority rawTenant:response.idTokenObj.realm context:nil error:nil];
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m]: done

File: IdentityCore/src/validation/MSIDAdfsAuthorityResolver.m [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+         {
+             if (completionBlock) completionBlock(response, error);
+             return;
+         }
+         
+         __auto_type drsCloudRequest = [[MSIDDRSDiscoveryRequest alloc] initWithDomain:domain adfsType:MSIDADFSTypeCloud];
+         drsCloudRequest.context = context;
+         [drsCloudRequest sendWithBlock:^(id response, NSError *error)
+          {
+              if (response)
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAdfsAuthorityResolver.m]: fixed

File: IdentityCore/src/validation/MSIDAADAuthority.m [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+    return authority;
+}
+
+#pragma mark - NSCopying
+
+- (id)copyWithZone:(NSZone *)zone
+{
+    MSIDAADAuthority *authority = [super copyWithZone:zone];
+    authority->_tenant = [_tenant copyWithZone:zone];
+    authority->_authorityCache = [MSIDAadAuthorityCache sharedInstance];
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAADAuthority.m]: fixed

File: IdentityCore/src/validation/MSIDAADAuthority.m [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+        return NO;
+    }
+    
+    return tenant != nil;
+}
+
++ (instancetype)aadAuthorityWithAuthorityURL:(NSURL *)authorityUrl
+                                   rawTenant:(NSString *)rawTenant
+                                     context:(id<MSIDRequestContext>)context
+                                       error:(NSError **)error;
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAADAuthority.m]: fixed

File: IdentityCore/src/validation/MSIDAADAuthority.m [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+{
+    if (![self.class isAuthorityFormatValid:authorityUrl context:context error:error]) return nil;
+    
+    authorityUrl = [NSURL URLWithString:[NSString stringWithFormat:@"https://%@/%@", [authorityUrl msidHostWithPortIfNecessary], rawTenant]];
+    
+    __auto_type authority = [[MSIDAADAuthority alloc] initWithURL:authorityUrl context:nil error:nil];
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAADAuthority.m]: done

File: IdentityCore/src/validation/MSIDAADAuthority.m [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+}
+
++ (instancetype)aadAuthorityWithAuthorityURL:(NSURL *)authorityUrl
+                                   rawTenant:(NSString *)rawTenant
+                                     context:(id<MSIDRequestContext>)context
+                                       error:(NSError **)error;
+{
+    if (![self.class isAuthorityFormatValid:authorityUrl context:context error:error]) return nil;
+    
+    authorityUrl = [NSURL URLWithString:[NSString stringWithFormat:@"https://%@/%@", [authorityUrl msidHostWithPortIfNecessary], rawTenant]];
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAADAuthority.m]: done

File: IdentityCore/src/validation/MSIDAADAuthority.m [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+            _tenant = [self.class tenantFromAuthorityUrl:self.url context:context error:error];
+        }
+    }
+    
+    return self;
+}
+
+- (void)setAuthorityCache:(MSIDAadAuthorityCache *)authorityCache
+{
+    _authorityCache = authorityCache ? authorityCache : [MSIDAadAuthorityCache sharedInstance];
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAADAuthority.m]: Removed it from public properties.

File: IdentityCore/src/validation/MSIDAADAuthority.m [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+
+- (instancetype)initWithURL:(NSURL *)url
+                    context:(id<MSIDRequestContext>)context
+                      error:(NSError **)error
+{
+    self = [super initWithURL:url context:context error:error];
+    if (self)
+    {
+        _url = [self.class normalizedAuthorityUrl:url context:context error:error];
+        if (!_url) return nil;
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAADAuthority.m]: it is created in line 39.

File: IdentityCore/src/oauth2/aad_base/MSIDAADOauth2Factory.m [hunk 13]
```diff
@@ -155,21 +163,35 @@ - (NSURL *)cacheURLForAuthority:(NSURL *)originalAuthority
...
+            
+            [aliases addObjectsFromArray:[aadAuthority cacheAliases]];
+        }
+        else
+        {
+            // If it's a tenanted authority, lookup original authority and common as those are the same, but start with original authority
+            [aliases addObjectsFromArray:[aadAuthority cacheAliases]];
+            
+            __auto_type aadAuthorityCommon = [MSIDAADAuthority aadAuthorityWithAuthorityURL:aadAuthority.url rawTenant:MSIDAADTenantTypeCommonRawValue context:nil error:nil];
+            [aliases addObjectsFromArray:[aadAuthorityCommon cacheAliases]];
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/oauth2/aad_base/MSIDAADOauth2Factory.m]: We shouldn't get any performance issues here, the performance is approximately same as it used to be.

File: IdentityCore/src/oauth2/aad_base/MSIDAADOauth2Factory.m [hunk 14]
```diff
@@ -143,9 +149,11 @@ - (NSURL *)cacheURLForAuthority:(NSURL *)originalAuthority
     {
         return nil;
     }
-
-    NSURL *authority = [MSIDAuthority universalAuthorityURL:originalAuthority];
-    return [[MSIDAadAuthorityCache sharedInstance] cacheUrlForAuthority:authority context:context];
+    
+    __auto_type authorityFactory = [MSIDAuthorityFactory new];
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/oauth2/aad_base/MSIDAADOauth2Factory.m]: removed `cacheURLForAuthority` since it is available in `authority` object.

File: IdentityCore/src/oauth2/aad_base/MSIDAADOauth2Factory.m [hunk 15]
```diff
@@ -128,7 +131,10 @@ - (NSString *)cacheEnvironmentFromEnvironment:(NSString *)originalEnvironment co
 
 - (NSArray<NSURL *> *)cacheAliasesForAuthority:(NSURL *)originalAuthority
 {
-    return [[MSIDAadAuthorityCache sharedInstance] cacheAliasesForAuthority:originalAuthority];
+    __auto_type authorityFactory = [MSIDAuthorityFactory new];
+    __auto_type authority = [authorityFactory authorityFromUrl:originalAuthority context:nil error:nil];
+    
+    return [authority cacheAliases] ?: @[];
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/oauth2/aad_base/MSIDAADOauth2Factory.m]: it could be nil when authority is nil

File: IdentityCore/src/network/request_configurator/MSIDAADRequestConfigurator.m [hunk 16]
```diff
@@ -60,13 +58,25 @@ - (void)configure:(MSIDHttpRequest *)request
...
+    __auto_type requestUrl = request.urlRequest.URL;
+    
+    __auto_type authorityFactory = [MSIDAuthorityFactory new];
+    __auto_type authority = [authorityFactory authorityFromUrl:request.urlRequest.URL context:request.context error:nil];
+    // If url is authority, then we are trying to get network url of it. Otherwise we use provided url.
+    __auto_type authorityUrl = [authority networkUrlWithContext:request.context];
+    
+    if (authorityUrl)
+    {
+        requestUrl = [requestUrl msidURLForPreferredHost:[authorityUrl msidHostWithPortIfNecessary] context:nil error:nil];
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/network/request_configurator/MSIDAADRequestConfigurator.m]: Network url has just host + port, like this:
`https://login.microsoftonline.com:80`

In NSURLRequest we have full url, like this: `https://login.microsoftonline.com:80/common/discovery/instance`

So what that logic does, it takes network host and replaces current host in NSURLRequest url.

File: IdentityCore/src/network/request/MSIDOpenIdConfigurationInfoRequest.m [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+    }
+    
+    __auto_type metadata = [MSIDOpenIdProviderMetadata new];
+    metadata.authorizationEndpoint = [NSURL URLWithString:jsonObject[@"authorization_endpoint"]];
+    metadata.tokenEndpoint = [NSURL URLWithString:jsonObject[@"token_endpoint"]];
+    
+    NSString *issuerString = jsonObject[@"issuer"];
+    
+    // If `issuer` contains {tenantid}, it is AAD authority.
+    // Lets exctract tenant from `endpoint` and put it instead of {tenantid}.
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/network/request/MSIDOpenIdConfigurationInfoRequest.m]: nope, it's still not fixed.

File: IdentityCore/src/network/MSIDAADEndpointProviding.h [hunk 18]
```diff
@@ -21,47 +21,22 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
-#import "MSIDAadAuthorityCache+TestUtil.h"
+#import <Foundation/Foundation.h>
+#import "MSIDADFSType.h"
 
-#include <pthread.h>
+@protocol MSIDAADEndpointProviding <NSObject>
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/network/MSIDAADEndpointProviding.h]: I suggest to keep it this way for now. We can separate it latter when we replace network requests with new one.

File: IdentityCore/src/oauth2/aad_base/MSIDAADOauth2Factory.m [hunk 19]
```diff
@@ -128,7 +131,10 @@ - (NSString *)cacheEnvironmentFromEnvironment:(NSString *)originalEnvironment co
 
 - (NSArray<NSURL *> *)cacheAliasesForAuthority:(NSURL *)originalAuthority
 {
-    return [[MSIDAadAuthorityCache sharedInstance] cacheAliasesForAuthority:originalAuthority];
+    __auto_type authorityFactory = [MSIDAuthorityFactory new];
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/oauth2/aad_base/MSIDAADOauth2Factory.m]: done

File: IdentityCore/src/oauth2/aad_base/MSIDAADOauth2Factory.m [hunk 20]
```diff
@@ -155,21 +163,35 @@ - (NSURL *)cacheURLForAuthority:(NSURL *)originalAuthority
...
 
-    NSMutableArray *lookupAuthorities = [NSMutableArray array];
-
-    if ([MSIDAuthority isTenantless:originalAuthority])
+    NSMutableArray *aliases = [NSMutableArray array];
+    
+    __auto_type authorityFactory = [MSIDAuthorityFactory new];
+    __auto_type authority = [authorityFactory authorityFromUrl:originalAuthority context:nil error:nil];
+    
+    if ([authority isKindOfClass:MSIDAADAuthority.class])
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/oauth2/aad_base/MSIDAADOauth2Factory.m]: done

File: IdentityCore/src/validation/MSIDAADAuthority.m [hunk 21]
```diff
@@ -0,0 +1,222 @@
...
+
+- (nullable instancetype)initWithURL:(nonnull NSURL *)url
+                           rawTenant:(NSString *)rawTenant
+                             context:(nullable id<MSIDRequestContext>)context
+                               error:(NSError **)error
+{
+    self = [self initWithURL:url context:context error:error];
+    if (self)
+    {
+        if (rawTenant && [self.tenant isTenantless])
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAADAuthority.m]: removed isTenantless

File: IdentityCore/src/validation/MSIDAADAuthority.m [hunk 22]
```diff
@@ -0,0 +1,222 @@
...
+
++ (BOOL)isAuthorityFormatValid:(NSURL *)url
+                       context:(id<MSIDRequestContext>)context
+                         error:(NSError **)error
+{
+    if (![super isAuthorityFormatValid:url context:context error:error]) return NO;
+    
+    __auto_type tenant = [self tenantFromAuthorityUrl:url context:context error:error];
+    
+    if ([tenant.rawTenant isEqualToString:@"adfs"])
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAADAuthority.m]: at least it will give a hint to developer that he is using wrong class, otherwise it would be possible to init AAD authority with ADFS url.

File: IdentityCore/src/validation/MSIDAADTenant.h [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+@interface MSIDAADTenant : NSObject <NSCopying>
+
+- (nullable instancetype)initWithRawTenant:(nonnull NSString *)rawTenant
+                             context:(nullable id<MSIDRequestContext>)context
+                               error:(NSError * _Nullable __autoreleasing * _Nullable)error;
+
+@property (nonatomic, readonly, nonnull) NSString *rawTenant;
+@property (nonatomic, readonly) MSIDAADTenantType type;
+
+- (BOOL)isTenantless;
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAADTenant.h]: removed it

File: IdentityCore/src/MSIDConstants.h [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+extern NSString * _Nonnull const MSID_PLATFORM_KEY;//The SDK platform. iOS or OSX
+extern NSString * _Nonnull const MSID_VERSION_KEY;
+extern NSString * _Nonnull const MSID_CPU_KEY;//E.g. ARM64
+extern NSString * _Nonnull const MSID_OS_VER_KEY;//iOS/OSX version
+extern NSString * _Nonnull const MSID_DEVICE_MODEL_KEY;//E.g. iPhone 5S
+
+extern NSString * _Nonnull const MSIDTrustedAuthority;
```
Comment (oldalton) [inline IdentityCore/src/MSIDConstants.h]: nit: can we follow the same format as other constants?

File: IdentityCore/src/network/request_configurator/MSIDAADRequestConfigurator.m (hunk 16 above)
[the remark this answers is not in the document]
Comment (oldalton) [reply IdentityCore/src/network/request_configurator/MSIDAADRequestConfigurator.m]: So, what it does, it tries to replace the host of request url with host from network url. Can we just query for network host and swap the hosts? Or at least add a more thorough comment. This logic is a bit confusing to read.

File: IdentityCore/src/validation/MSIDAADAuthority.m [hunk 25]
```diff
@@ -0,0 +1,247 @@
...
+
+- (NSURL *)networkUrlWithContext:(id<MSIDRequestContext>)context
+{
+    return [self.authorityCache networkUrlForAuthority:self context:context];
+}
+
+- (NSURL *)cacheUrlWithContext:(id<MSIDRequestContext>)context
+{
+    __auto_type universalAuthorityURL = [self universalAuthorityURL];
+    __auto_type authority = (MSIDAADAuthority *)[self.authorityFactory authorityFromUrl:universalAuthorityURL context:context error:nil];
```
Comment (oldalton) [inline IdentityCore/src/validation/MSIDAADAuthority.m]: nit: why are we ignoring error?

File: IdentityCore/src/validation/MSIDAADAuthority.m [hunk 26]
```diff
@@ -0,0 +1,247 @@
...
+{
+    __auto_type authorityUrl = [NSURL URLWithString:[NSString stringWithFormat:@"https://%@/%@", environment, rawTenant]];
+    __auto_type authority = [[MSIDAADAuthority alloc] initWithURL:authorityUrl context:context error:error];
+    
+    return authority;
+}
+
+- (nonnull NSString *)authorityType
+{
+    return MSID_TELEMETRY_VALUE_AUTHORITY_AAD;
```
Comment (oldalton) [inline IdentityCore/src/validation/MSIDAADAuthority.m]: nit: it's a bit weird to use a telemetry constant for authority type here. Can we rename the constant or rename the method to telemetryAuthorityType?

File: IdentityCore/src/validation/MSIDAADAuthority.m [hunk 27]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+#pragma mark - NSCopying
+
+- (id)copyWithZone:(NSZone *)zone
+{
+    MSIDAADAuthority *authority = [super copyWithZone:zone];
+    authority->_tenant = [_tenant copyWithZone:zone];
```
Comment (oldalton) [inline IdentityCore/src/validation/MSIDAADAuthority.m]: should we also copy openIdConfigurationEndpoint?

File: IdentityCore/src/validation/MSIDAadAuthorityCache.m [hunk 28]
```diff
@@ -74,63 +53,51 @@ + (MSIDAadAuthorityCache *)sharedInstance
...
-    {
-        // Run this through urlForPreferredHost to make sure it does not throw any exceptions
-        urlForPreferredHost([NSURL URLWithString:@"https://fakeurl.contoso.com"], host);
-        
-        return YES;
-    }
-    @catch (NSException *ex)
+    // Run this through urlForPreferredHost to make sure it does not return any errors.
+    NSError *err;
+    [[NSURL URLWithString:@"https://fakeurl.contoso.com"] msidURLForPreferredHost:host context:context error:&err];
```
Comment (oldalton) [inline IdentityCore/src/validation/MSIDAadAuthorityCache.m]: why did you remove the try/catch from here?

File: IdentityCore/src/validation/MSIDAuthority.h [hunk 29]
```diff
@@ -17,57 +15,75 @@
...
+
+- (instancetype _Nullable )init NS_UNAVAILABLE;
++ (instancetype _Nullable )new NS_UNAVAILABLE;
+
+- (nullable instancetype)initWithURL:(nonnull NSURL *)url
+                             context:(nullable id<MSIDRequestContext>)context
+                               error:(NSError * _Nullable __autoreleasing * _Nullable)error NS_DESIGNATED_INITIALIZER;
+
+- (nullable instancetype)initWithURL:(nonnull NSURL *)url
+                           rawTenant:(nullable NSString *)rawTenant
```
Comment (oldalton) [inline IdentityCore/src/validation/MSIDAuthority.h]: Do we need initializer with tenant in the base class?

File: IdentityCore/src/validation/MSIDAuthority.h [hunk 30]
```diff
@@ -17,57 +15,75 @@
...
+- (nonnull NSURL *)networkUrlWithContext:(nullable id<MSIDRequestContext>)context;
 
-@property (class, readonly, nonnull) NSCache *openIdConfigurationCache;
+- (nonnull NSURL *)cacheUrlWithContext:(nullable id<MSIDRequestContext>)context;
 
-+ (BOOL)isADFSInstance:(nonnull NSString *)endpoint;
-+ (BOOL)isADFSInstanceURL:(nonnull NSURL *)endpointUrl;
-+ (BOOL)isConsumerInstanceURL:(nonnull NSURL *)authorityURL;
-+ (BOOL)isB2CInstanceURL:(nonnull NSURL *)endpointUrl;
+- (nonnull NSArray<NSURL *> *)cacheAliases;
```
Comment (oldalton) [inline IdentityCore/src/validation/MSIDAuthority.h]: In the new cache we don't operate based on alias URLs, we operate based on environments which are strings. How can we now get a list of environment aliases?

File: IdentityCore/src/validation/MSIDAuthorityFactory.m [hunk 31]
```diff
@@ -0,0 +1,96 @@
...
+    
+    if ([MSIDAADAuthority isAuthorityFormatValid:url context:context error:nil])
+    {
+        __auto_type aadAuthority = [[MSIDAADAuthority alloc] initWithURL:url context:context error:nil];
+        if (aadAuthority) return aadAuthority;
+    }
+    
+    if (error)
+    {
+        *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInvalidDeveloperParameter, @"Provided authority url is not a valid authority.", nil, nil, nil, context.correlationId, nil);
```
Comment (oldalton) [inline IdentityCore/src/validation/MSIDAuthorityFactory.m]: Should we create a base OIDC authority instead?
Let's create an issue to follow up... (don't block on this)

File: IdentityCore/src/validation/MSIDAuthorityFactory.m [hunk 32]
```diff
@@ -0,0 +1,96 @@
...
+                            context:(id<MSIDRequestContext>)context
+                              error:(NSError **)error
+{
+    if ([MSIDB2CAuthority isAuthorityFormatValid:url context:context error:nil])
+    {
+        __auto_type b2cAuthority = [[MSIDB2CAuthority alloc] initWithURL:url rawTenant:rawTenant context:context error:nil];
+        if (b2cAuthority) return b2cAuthority;
+    }
+    
+    if ([MSIDAADAuthority isAuthorityFormatValid:url context:context error:nil])
```
Comment (oldalton) [inline IdentityCore/src/validation/MSIDAuthorityFactory.m]: this checks for MSIDAADAuthority and creates ADFS authority

File: IdentityCore/src/validation/MSIDAuthorityFactory.m [hunk 33]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+- (MSIDAuthority *)authorityFromUrl:(NSURL *)url
+                          rawTenant:(NSString *)rawTenant
+                            context:(id<MSIDRequestContext>)context
+                              error:(NSError **)error
+{
+    if ([MSIDB2CAuthority isAuthorityFormatValid:url context:context error:nil])
+    {
+        __auto_type b2cAuthority = [[MSIDB2CAuthority alloc] initWithURL:url rawTenant:rawTenant context:context error:nil];
```
Comment (oldalton) [inline IdentityCore/src/validation/MSIDAuthorityFactory.m]: none of the initializers takes error?

File: IdentityCore/src/validation/MSIDAuthorityFactory.m [hunk 34]
```diff
@@ -0,0 +1,96 @@
...
+#import "MSIDADFSAuthority.h"
+#import "MSIDB2CAuthority.h"
+
+@implementation MSIDAuthorityFactory
+
+- (MSIDAuthority *)authorityFromUrl:(NSURL *)url
+                               context:(id<MSIDRequestContext>)context
+                                 error:(NSError **)error
+{
+    if ([MSIDB2CAuthority isAuthorityFormatValid:url context:context error:nil])
```
Comment (oldalton) [inline IdentityCore/src/validation/MSIDAuthorityFactory.m]: this code looks quite similar to the next one except there's no tenant. Can we just pass nil tenant to initializer?

File: IdentityCore/src/MSIDAADNetworkConfiguration.m [hunk 35]
```diff
@@ -33,6 +35,14 @@ + (void)initialize
     if (self == [MSIDAADNetworkConfiguration self])
     {
         s_defaultConfiguration = [MSIDAADNetworkConfiguration new];
+        
+        s_trustedHostList = [NSSet setWithObjects:MSIDTrustedAuthority,
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/MSIDAADNetworkConfiguration.m]: done

File: IdentityCore/src/validation/MSIDAADAuthority.m (hunk 26 above)
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAADAuthority.m]: renamed to telemetryAuthorityType

File: IdentityCore/src/validation/MSIDAADAuthority.m (hunk 27 above)
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAADAuthority.m]: done

File: IdentityCore/src/validation/MSIDAadAuthorityCache.m (hunk 28 above)
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAadAuthorityCache.m]: try/catch moved to msidURLForPreferredHost

File: IdentityCore/src/validation/MSIDAuthority.h (hunk 29 above)
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAuthority.h]: Removed it from MSIDAuthority.

File: IdentityCore/src/validation/MSIDAuthority.h (hunk 30 above)
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAuthority.h]: We still can get them from MSIDOauth2Factory. On the other hand, I don't think that MSIDOauth2Factory is a right place for such logic, that is why I suggest to revise it latter, see #222.

File: IdentityCore/src/validation/MSIDAuthorityFactory.m (hunk 31 above)
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAuthorityFactory.m]: Probably we should. Agree, lets create issue for it.

File: IdentityCore/src/validation/MSIDAuthorityFactory.m (hunk 32 above)
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAuthorityFactory.m]: fixed

File: IdentityCore/src/validation/MSIDAuthorityFactory.m (hunk 34 above)
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAuthorityFactory.m]: good point, refactored.

File: IdentityCore/src/MSIDCache.m [hunk 36]
```diff
@@ -0,0 +1,100 @@
...
+        object = [self.container objectForKey:key];
+    });
+    
+    return object;
+}
+
+- (void)setObject:(id)obj forKey:(id)key
+{
+    dispatch_barrier_sync(self.synchronizationQueue, ^{
+        [self.container setObject:obj forKey:key];
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/MSIDCache.m]: yes, it was fixed in another PR.

File: IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m [hunk 37]
```diff
@@ -600,8 +602,8 @@ - (BOOL)saveAccessToken:(MSIDAccessToken *)accessToken
     // Delete access tokens with intersecting scopes
     MSIDDefaultCredentialCacheQuery *query = [MSIDDefaultCredentialCacheQuery new];
     query.homeAccountId = accessToken.homeAccountId;
-    query.environment = accessToken.authority.msidHostWithPortIfNecessary;
-    query.realm = accessToken.authority.msidTenant;
+    query.environment = accessToken.authority.url.msidHostWithPortIfNecessary;
+    query.realm = accessToken.authority.url.msidTenant;
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m]: We have tenant in MSIDAADAuthority, but since `authority` here is MSIDAuthority, we can not get tenant without dynamic cast to MSIDAADAuthority.

File: IdentityCore/src/validation/MSIDAuthorityFactory.m (hunk 33 above)
Comment (antrix1989) [reply IdentityCore/src/validation/MSIDAuthorityFactory.m]: Added underlyingError.

File: IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m [hunk 38]
```diff
@@ -229,8 +231,8 @@ - (MSIDAccessToken *)getAccessTokenForAccount:(MSIDAccountIdentifier *)account
 
     MSIDDefaultCredentialCacheQuery *query = [MSIDDefaultCredentialCacheQuery new];
     query.homeAccountId = account.homeAccountId;
-    query.environment = configuration.authority.msidHostWithPortIfNecessary;
-    query.realm = configuration.authority.msidTenant;
+    query.environment = configuration.authority.url.msidHostWithPortIfNecessary;
```
[the remark this answers is not in the document]
Comment (antrix1989) [reply IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m]: done

File: IdentityCore/src/MSIDConstants.h (hunk 24 above)
Comment (antrix1989) [reply IdentityCore/src/MSIDConstants.h]: Decided to keep camel case constants and change upper case constants to camel case latter.

File: IdentityCore/src/validation/MSIDAADAuthority.m (hunk 27 above)
Comment (oldalton) [reply IdentityCore/src/validation/MSIDAADAuthority.m]: did it get done?
